### ISSUE TITLE
plugin can't handle objects that can be null

### ISSUE THREAD

**Issue by metzc** [angular/vscode-ng-language-service#71]
When i have this type on my component:

```
    test: {a: number, b: number} | null = {
        a: 1,
        b: 2
    }
```

and this in the template:
```
{{test?.a}}
```

the plugin complains and tells me:
```
'Identifier 'a' is not defined. '<anonymous>' does not contain such a member'
```

I use the `?.` accessor method all the time (of course one could do `*ngIf="test"` on the outer element but it turns out this does not solve the problem. `test` is still treated as `<anonymous>` in that case (along with the error message), meaning that the the `?.` operator is not the problem here). Seemingly the plugin can't handle access to members of objects that are possibly `null`. If i remove `null` from the type definition everything works smoothly (except the rest of code breaking ;-).

**Comment by chuckjaz** [comment]
I have reproduced this.

Note that this requires "strictNullChecks" to be enabled in the `tsconfig.json` file.

Tracking in Angular: https://github.com/angular/angular/issues/15885

**Comment by metzc** [comment]
Please note that this also happens when wrapping the expression with an ngIf, so this issue is not only related to the ?. operator.

**Comment by chuckjaz** [comment]
The `*ngIf` does not affect the type of the expression. The language service does not treat common directives specially and there is no way currently to communicate that `*ngIf` acts as a guard for the referenced template.

The reason we don't threat `*ngIf` as special is because we want third-parties to be able to create equivalent directives. In other words, if I treated `*ngIf` special then no else could create the equivalent of `*ngIf`. To avoid this I need a declarative way to communicate that in the template referenced by `ngThen` the expression bound to `ngIf` is truthy and the expression bound to `ngIf` in the `ngIfElse` is always falsy. We currently have no way to describe this.

This will not be addressed until we change how we do type inferencing in the compiler which I am working on a proposal for now.

**Comment by chuckjaz** [comment]
What I plan on implementing, probably for 4.1, is support for `!` which would meant assume the preceding expression is not `null` or `undefined`. We would need this even if we implement the above.

**Comment by chuckjaz** [comment]
Fixed in 0.1.4
